### PR TITLE
#1502: removed any types in auth.utils.ts

### LINKED ISSUE #1502: [Refactoring] - Remove `any` type from `src/backend/src/utils/auth.utils.ts`
### Description

Remove the use of the `any` type from auth.utils.ts This may seem very simple, but the point of this is just to have you read through the codebase and understand the types/interfaces used in it, or make new ones that fit the shape of the data being used.

### Notes

Check #665 for notes on why we're doing this and pointers on what you should replace `any` with.

### Acceptance Criteria

`any` is not used as a type in auth.utils.ts

### PR DESCRIPTION
**Pull request by chpy04**
## Changes

removed any types in auth.utils.ts using the types NextFunction from express and JwtPayload and VerifyErrors from 'jsonwebtoken'.

## Notes
uses the VerifyCallback type in the index.d.ts file for err and decoded types in the jwt.verify function call.
the (err: any, decoded: any) argument is a VerifyCallback which means the decoded can be a string | JwtPayload | undefined, but only the JwtPayload would have a userId, so I threw an error if it wasn't a JwtPayload.

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #1502

### CODE REVIEW

**Review by jackturner106 — DISMISSED** [review]
Looks Good to Me!

File: src/backend/src/utils/auth.utils.ts [hunk 1]
```diff
@@ -36,9 +37,12 @@ export const requireJwtProd = (req: Request, res: Response, next: any) => {
 
     if (!token) return res.status(401).json({ message: 'Authentication Failed: Cookie not found!' });
 
-    jwt.verify(token, TOKEN_SECRET, (err: any, decoded: any) => {
+    jwt.verify(token, TOKEN_SECRET, (err: VerifyErrors | null, decoded: string | JwtPayload | undefined) => {
       if (err) return res.status(401).json({ message: 'Authentication Failed: Invalid JWT!' });
 
+      if (typeof decoded === 'undefined' || typeof decoded === 'string') {
```
Comment (RChandler234) [inline src/backend/src/utils/auth.utils.ts:43]: you should be able to just do !decoded || here

**Review by RChandler234 — CHANGES_REQUESTED** [review]
Can u also document how you got these types a lil bit, this stuff is messing with auth, so I wanna triple check that it's right

File: src/backend/src/utils/auth.utils.ts [hunk 2]
```diff
@@ -36,9 +37,14 @@ export const requireJwtProd = (req: Request, res: Response, next: any) => {
 
     if (!token) return res.status(401).json({ message: 'Authentication Failed: Cookie not found!' });
 
-    jwt.verify(token, TOKEN_SECRET, (err: any, decoded: any) => {
+    //uses the VerifyCallback type in the index.d.ts file for err and decoded types.
+    //Only jwtPayload of the decoded types has a userID, so other types throw an error
```
Comment (RChandler234) [inline src/backend/src/utils/auth.utils.ts:41]: Sorry, I meant put these comments on the PR for our reference, these are probably a bit too detailed for comments in the code

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: src/backend/src/utils/auth.utils.ts [hunk 3]
```diff
@@ -36,9 +37,12 @@ export const requireJwtProd = (req: Request, res: Response, next: any) => {
 
     if (!token) return res.status(401).json({ message: 'Authentication Failed: Cookie not found!' });
 
-    jwt.verify(token, TOKEN_SECRET, (err: any, decoded: any) => {
+    jwt.verify(token, TOKEN_SECRET, (err: VerifyErrors | null, decoded: string | JwtPayload | undefined) => {
```
Comment (RChandler234) [inline src/backend/src/utils/auth.utils.ts:40]: I beleive this should be `string | JwtPayload | Jwt` based on the dependency file, is there a specific reason you left it out?

**Review by RChandler234 — CHANGES_REQUESTED** [review]
One more comment, sorry I know this one is tedious, but I need to be thorough since it's related to login/ authentication

File: src/backend/src/utils/auth.utils.ts (hunk 3 above)
Comment (chpy04) [reply src/backend/src/utils/auth.utils.ts:40]: Of the four asynchronous verify functions in the index.d.ts file, only the first one takes 3 parameters (on line 215). The other four take in an VerifyOptions as well as the token, key, and VerifyCallback. In the function with 3 parameters, the callback parameter is typed as VerifyCallback<JwtPayload | string>. It was my understanding that the JwtPayload | string becomes the T in the VerifyCallback type, which is why I didn't include the Jwt in the decoded type. That being said, I can add the Jwt type in and check for it the same as the string if you think that would be better, or I can explain the stuff I just explained here in the pr if you think it makes sense.

**Review by RChandler234 — APPROVED** [review]
LGTM 🔥

**Pull request merged**